api/src/main/java/bio/terra/tanagra/service/query/api/FilterConverter.java

```diff
@@ -0,0 +1,139 @@
+package bio.terra.tanagra.service.query.api;
+
+import bio.terra.common.exception.BadRequestException;
+import bio.terra.tanagra.generated.model.ApiArrayFilter;
+import bio.terra.tanagra.generated.model.ApiArrayFilterOperator;
+import bio.terra.tanagra.generated.model.ApiBinaryFilter;
+import bio.terra.tanagra.generated.model.ApiBinaryFilterOperator;
+import bio.terra.tanagra.generated.model.ApiFilter;
+import bio.terra.tanagra.generated.model.ApiRelationshipFilter;
+import bio.terra.tanagra.service.search.Entity;
+import bio.terra.tanagra.service.search.EntityVariable;
+import bio.terra.tanagra.service.search.Expression.AttributeExpression;
+import bio.terra.tanagra.service.search.Expression.Literal;
+import bio.terra.tanagra.service.search.Filter;
+import bio.terra.tanagra.service.underlay.Underlay;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.util.Objects;
+import java.util.Optional;
+
+/** Converts API filters to Tanagra search {@link bio.terra.tanagra.service.search.Filter}s. */
+class FilterConverter {
+  private final Underlay underlay;
+  private final ExpressionConverter expressionConverter;
+
+  FilterConverter(Underlay underlay) {
+    this.underlay = underlay;
+    this.expressionConverter = new ExpressionConverter(underlay);
+  }
+
+  public Filter convert(ApiFilter apiFilter, VariableScope scope) {
+    if (!ConversionUtils.exactlyOneNonNull(
+        apiFilter.getArrayFilter(),
+        apiFilter.getBinaryFilter(),
+        apiFilter.getRelationshipFilter())) {
+      throw new BadRequestException(
+          "Filter must have exactly one non-null field.\n" + apiFilter.toString());
+    }
+    if (apiFilter.getArrayFilter() != null) {
+      return convert(apiFilter.getArrayFilter(), scope);
+    } else if (apiFilter.getBinaryFilter() != null) {
+      return convert(apiFilter.getBinaryFilter(), scope);
+    } else if (apiFilter.getRelationshipFilter() != null) {
+      return convert(apiFilter.getRelationshipFilter(), scope);
+    }
+    throw new BadRequestException("Unknown Filter type: " + apiFilter.toString());
+  }
+
+  @VisibleForTesting
+  Filter.BinaryFunction convert(ApiBinaryFilter apiBinary, VariableScope scope) {
+    Preconditions.checkNotNull(
```

Comment:
should these be `BadRequestException` as well? This is basically request validation which should return a 400 - right?

Reply:
Yep, thanks.